### ISSUE TITLE
AIoE stopped publishing on MQTT

### ISSUE THREAD

**Issue by EleRas** [jomjol/AI-on-the-edge-device#1856]
### The Problem

While trying to detect digits it seems that AIoE disconnects from the MQTT server and does not publish any data. It tries twice to publish, gives up, reconnects and publishes all metadata. No values are exported via MQTT. The MQTT server ran stable during this time, but reported a new connection by the client and closing the old one.
Notice the lines at the following times, there seem to be multiple disconnects without any new connection?
```
[0d00h10m32s] 2023-01-16T20:17:59 <DBG> [MQTT IF] Disconnected from broker
[0d00h10m47s] 2023-01-16T20:18:14 <DBG> [MQTT IF] Disconnected from broker
[0d00h11m02s] 2023-01-16T20:18:29 <DBG> [MQTT IF] Disconnected from broker
```

### Version

rolling (Commit: 22ef80d+)

### Logfile

```shell
[0d00h10m14s] 2023-01-16T20:17:41 <DBG> [TFLITE SERVER] ----------------------------------------------------------------
[0d00h10m14s] 2023-01-16T20:17:41 <INF> [TFLITE SERVER] Round #3 started
[0d00h10m14s] 2023-01-16T20:17:41 <DBG> [MQTT IF] Published topic: gaszaehler/status, content: Take Image (msg_id=114)
[0d00h10m23s] 2023-01-16T20:17:51 <DBG> [MQTT IF] Published topic: gaszaehler/status, content: Aligning (msg_id=115)
[0d00h10m32s] 2023-01-16T20:17:59 <DBG> [MQTT IF] Disconnected from broker
[0d00h10m34s] 2023-01-16T20:18:02 <DBG> [MQTT IF] Publish skipped. Client not initalized or not connected. (topic: gaszaehler/status)
[0d00h10m34s] 2023-01-16T20:18:02 <DBG> [CNN] doFlow after alignment
[0d00h10m34s] 2023-01-16T20:18:02 <DBG> [TFLITE] CTfLiteClass::LoadModel
[0d00h10m34s] 2023-01-16T20:18:02 <DBG> [TFLITE] CTfLiteClass::ReadFileToModel: /sdcard/config/dig-cont_0600_s3.tflite
[0d00h10m37s] 2023-01-16T20:18:04 <DBG> [TFLITE] CTfLiteClass::MakeAllocate
[0d00h10m37s] 2023-01-16T20:18:04 <DBG> [CNN] Processing Number 'main'
[0d00h10m37s] 2023-01-16T20:18:04 <DBG> [CNN] ROI #0 - TfLite
[0d00h10m37s] 2023-01-16T20:18:04 <DBG> [CNN] CNN Type: DoubleHyprid10
[0d00h10m42s] 2023-01-16T20:18:09 <DBG> [CNN] After Invoke
[0d00h10m42s] 2023-01-16T20:18:09 <DBG> [CNN] _num (p, m): 3 4 2 _val (p, m): 0.792597 0.077307 0.008673 result: 3.088868 _fit: 0.869904
[0d00h10m42s] 2023-01-16T20:18:09 <DBG> [CNN] ROI #1 - TfLite
[0d00h10m42s] 2023-01-16T20:18:09 <DBG> [CNN] CNN Type: DoubleHyprid10
[0d00h10m47s] 2023-01-16T20:18:14 <DBG> [CNN] After Invoke
[0d00h10m47s] 2023-01-16T20:18:14 <DBG> [CNN] _num (p, m): 4 5 3 _val (p, m): 0.805752 0.097417 0.012502 result: 4.107861 _fit: 0.903168
[0d00h10m47s] 2023-01-16T20:18:14 <DBG> [CNN] ROI #2 - TfLite
[0d00h10m47s] 2023-01-16T20:18:14 <DBG> [CNN] CNN Type: DoubleHyprid10
[0d00h10m47s] 2023-01-16T20:18:14 <DBG> [MQTT IF] Disconnected from broker
[0d00h10m52s] 2023-01-16T20:18:19 <DBG> [CNN] After Invoke
[0d00h10m52s] 2023-01-16T20:18:19 <DBG> [CNN] _num (p, m): 7 8 6 _val (p, m): 0.982209 0.002298 0.015259 result: 6.984702 _fit: 0.997468
[0d00h10m52s] 2023-01-16T20:18:19 <DBG> [CNN] ROI #3 - TfLite
[0d00h10m52s] 2023-01-16T20:18:19 <DBG> [CNN] CNN Type: DoubleHyprid10
[0d00h10m57s] 2023-01-16T20:18:24 <DBG> [CNN] After Invoke
[0d00h10m57s] 2023-01-16T20:18:24 <DBG> [CNN] _num (p, m): 2 3 1 _val (p, m): 0.963136 0.000194 0.036637 result: 1.963355 _fit: 0.999773
[0d00h10m57s] 2023-01-16T20:18:24 <DBG> [CNN] ROI #4 - TfLite
[0d00h10m57s] 2023-01-16T20:18:24 <DBG> [CNN] CNN Type: DoubleHyprid10
[0d00h11m02s] 2023-01-16T20:18:29 <DBG> [CNN] After Invoke
[0d00h11m02s] 2023-01-16T20:18:29 <DBG> [CNN] _num (p, m): 6 7 5 _val (p, m): 0.985673 0.000004 0.014313 result: 5.985687 _fit: 0.999986
[0d00h11m02s] 2023-01-16T20:18:29 <DBG> [CNN] ROI #5 - TfLite
[0d00h11m02s] 2023-01-16T20:18:29 <DBG> [CNN] CNN Type: DoubleHyprid10
[0d00h11m02s] 2023-01-16T20:18:29 <DBG> [MQTT IF] Disconnected from broker
[0d00h11m07s] 2023-01-16T20:18:34 <DBG> [CNN] After Invoke
[0d00h11m07s] 2023-01-16T20:18:34 <DBG> [CNN] _num (p, m): 9 0 8 _val (p, m): 0.950120 0.043048 0.000118 result: 9.043344 _fit: 0.993169
[0d00h11m07s] 2023-01-16T20:18:34 <DBG> [CNN] ROI #6 - TfLite
[0d00h11m07s] 2023-01-16T20:18:34 <DBG> [CNN] CNN Type: DoubleHyprid10
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] After Invoke
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] _num (p, m): 7 8 6 _val (p, m): 0.994586 0.001913 0.003479 result: 6.996514 _fit: 0.998065
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [MQTT IF] Publish skipped. Client not initalized or not connected. (topic: gaszaehler/status)
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout _analog=0, _extendedResolution=0, prev=-1
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] PointerEvalHybridNew - No predecessor - Result = 7 number: 6.996514 number_of_predecessors = -1.000000 eval_predecessors = -1 Digital_Uncertainty = 0.200000
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout(dig100) prev=7
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] PointerEvalHybridNew - NO analogue predecessor, no change of digits, as pre-decimal point far enough away = 9 number: 9.043344 number_of_predecessors = 6.996514 eval_predecessors = 7 Digital_Uncertainty = 0.200000
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#PointerEvalHybridNew()= 9
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#result= 97
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] PointerEvalHybridNew - NO analogue predecessor, no change of digits, as pre-decimal point far enough away = 6 number: 5.985687 number_of_predecessors = 9.043344 eval_predecessors = 9 Digital_Uncertainty = 0.200000
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#PointerEvalHybridNew()= 6
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#result= 697
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] PointerEvalHybridNew - NO analogue predecessor, no change of digits, as pre-decimal point far enough away = 2 number: 1.963355 number_of_predecessors = 5.985687 eval_predecessors = 6 Digital_Uncertainty = 0.200000
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#PointerEvalHybridNew()= 2
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#result= 2697
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] PointerEvalHybridNew - NO analogue predecessor, no change of digits, as pre-decimal point far enough away = 7 number: 6.984702 number_of_predecessors = 1.963355 eval_predecessors = 2 Digital_Uncertainty = 0.200000
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#PointerEvalHybridNew()= 7
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#result= 72697
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] PointerEvalHybridNew - NO analogue predecessor, no change of digits, as pre-decimal point far enough away = 4 number: 4.107861 number_of_predecessors = 6.984702 eval_predecessors = 7 Digital_Uncertainty = 0.200000
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#PointerEvalHybridNew()= 4
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#result= 472697
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] PointerEvalHybridNew - NO analogue predecessor, no change of digits, as pre-decimal point far enough away = 3 number: 3.088868 number_of_predecessors = 4.107861 eval_predecessors = 4 Digital_Uncertainty = 0.200000
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#PointerEvalHybridNew()= 3
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [CNN] getReadout#result= 3472697
[0d00h11m12s] 2023-01-16T20:18:39 <INF> [POSTPROC] main: Raw: 34726.97, Value: 34726.97, Status: no error
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [MQTT IF] Publish skipped. Client not initalized or not connected. (topic: gaszaehler/status)
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [MQTT IF] Publish skipped. Client not initalized or not connected. (topic: gaszaehler/status)
[0d00h11m12s] 2023-01-16T20:18:39 <DBG> [TFLITE SERVER] CPU Temperature: 57.2
[0d00h11m12s] 2023-01-16T20:18:39 <INF> [TFLITE SERVER] Round #3 completed (58 seconds)
[0d00h11m17s] 2023-01-16T20:18:44 <INF> [MQTT IF] Connected to broker
[0d00h11m17s] 2023-01-16T20:18:44 <DBG> [MQTT IF] Published topic: gaszaehler/connection, content: connected (msg_id=116)
[0d00h11m17s] 2023-01-16T20:18:44 <DBG> [MQTT IF] topic gaszaehler/ctrl/flow_start subscribe successful, msg_id=117
[0d00h11m19s] 2023-01-16T20:18:46 <DBG> [MAIN SERVER] info_get_handler
[0d00h11m25s] 2023-01-16T20:18:52 <DBG> [OTA FILE] log_get_last_part_handler
```

**Comment by caco3** [comment]
Yes, the MQTT client is not 100% stable, but we do not understand why.
In my tests, it succeeded again on the next round. How was it on your side?

**Comment by Slider0007** [comment]
What about your WIFI connection. Do you have stable connection? RSSI value?

It's interessting to see that during flow is processing the system is not able to reconnect, though. Does this happen every round or only sporadically?

I only have issues whenever the WIFI connection is quite bad, but it reconnects every time as soon as WIFI connection is stable enough again.

> Notice the lines at the following times, there seem to be multiple disconnects without any new connection?
> 
> ```
> [0d00h10m32s] 2023-01-16T20:17:59 <DBG> [MQTT IF] Disconnected from broker
> [0d00h10m47s] 2023-01-16T20:18:14 <DBG> [MQTT IF] Disconnected from broker
> [0d00h11m02s] 2023-01-16T20:18:29 <DBG> [MQTT IF] Disconnected from broker
> ```

The system is trying to reconnect every 15s. This message indicates only the actual state which means that the system is still not connected to the broker. Actually, this is no disconnection action itself.

**Comment by EleRas** [comment]
Sorry, for the delayed response: It sometimes takes multiple hours until the next MQTT update is available.
The Wifi connection is stable (AP does not show any disconnects, also no disconnects visible in the device logfiles).

I'm also confused that it says: "Connected to broker" & "Published topic: gaszaehler/connection, content: connected (msg_id=116)" (I verified that it updates on Mosquitto), but as soon as it tries to publish the actual data it says "Disconnected"

**Comment by Slider0007** [comment]
@EleRas: I just realized you are using the dig-cont_0600_s3 model.
`[0d00h10m34s] 2023-01-16T20:18:02 <DBG> [TFLITE] CTfLiteClass::ReadFileToModel: /sdcard/config/dig-cont_0600_s3.tflite`

In v13.0.8 there are some memory constraints when using large tflite models like dig-cont_0600_s3. To verify that your issue is is linked with too low memory you could temporarily change to a model which has smaller file size like one of the dig-class100-0xxx models.

**Comment by EleRas** [comment]
I changed to dig-class100-0150_s2_q.tflite - but sadly my experience tells me: this one basically never gets the digits correct

**Comment by Slider0007** [comment]
This change of the model can be just temporarily for testing purpose to figure out if this solves your mqtt issue.
If this solves it, issue is related to memory. With next official version using dig_con-600 would be an issue anymore, because some memory improvements are implemented. Either you wait for the next official release or you could test latest dev version using the link provided in next post. 
If it's not solves your issue, the issue is related to something else...

**Comment by EleRas** [comment]
Since the ESP was already on rolling, but a version which does not seem to contain the memory fixes, I updated to commit #287302b and activated dig_con-0600_s3.tflite.
Will report back in a few days if this fixes the issue.

**Comment by EleRas** [comment]
It seems to work now, closing the ticket for now and if it re-appears I'll reopen
